src/library/scala/collection/immutable/OrderedMap.scala

```diff
@@ -0,0 +1,518 @@
+        while (rearIter.hasNext) {
+          mapping = mapping - rearIter.next()
+        }
+        new OrderedMap(ordering, mapping, ordinal, orderedBy)
+      }
+    }
+  }
+
+  def get(key: K): Option[V] = mapping.get(key).map(value)
+
+  def iterator: Iterator[(K, V)] = new AbstractIterator[(K, V)] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): (K, V) = binding(iter.next())
+  }
+
+  override def keysIterator: Iterator[K] = new AbstractIterator[K] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): K = iter.next()
+  }
+
+  override def valuesIterator: Iterator[V] = new AbstractIterator[V] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): V = value(binding(iter.next()))
+  }
+
+  override def concat[V1 >: V](suffix: IterableOnce[(K, V1)]): OrderedMap[K, V1] = {
+    var result: OrderedMap[K, V1] = this
+    val iter = suffix.iterator
+    while (iter.hasNext) {
+      val (k, v) = iter.next()
+      result = result.updated(k, v)
+    }
+    result
+  }
+
+  override def map[K2, V2](f: ((K, V)) => (K2, V2)): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.map {
+      case (k, (o, v)) =>
+        val (k2, v2) = f((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  override def flatMap[K2, V2](f: ((K, V)) => IterableOnce[(K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    var mng: Mapping[K2, V2] = Mapping.empty
+    var ord = 0
+    val i = ordering.iterator
+    while (i.hasNext) {
+      val k = i.next()
+      val (o, v) = mapping(k)
+      val j = f((k, v)).iterator
+      while (j.hasNext) {
+        val (k2, v2) = j.next()
+        if ((orderedBy == OrderBy.Modification && k != k2 && v != v2) || !mng.contains(k2)) {
+          mng += k2 -> (ord, v2)
+          ong = ong.incl(ord, k2)
+        }
+        ord += 1
+      }
+    }
+    new OrderedMap[K2, V2](ong, mng, ord, orderedBy)
+  }
+
+  override def collect[K2, V2](pf: PartialFunction[(K, V), (K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.collect {
+      case (k, (o, v)) if pf.isDefinedAt((k, v)) =>
+        val (k2, v2) = pf((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  @`inline` private[this] def value(p: (_, V)) = p._2
+  @`inline` private[this] def binding(k: K) = mapping(k).copy(_1 = k)
+}
+object OrderedMap extends MapFactory[OrderedMap] {
+  sealed trait OrderBy
+  final object OrderBy {
+    final case object Insertion extends OrderBy
+    final case object Modification extends OrderBy
+  }
```

Comment:
While a user can use the remove-insert method for explicit calls to get modification order, as it is now the ordering also affects implicit reinsertions (such as those that occur during transformation calls (`map`/`collect`/`flatMap`) when the transformation function returns identical keys for different entries. For this reason I have kept the `orderingBy`-feature for now but it would be easy to remove it if so required).